contracts/contracts/vault/VaultAdmin.sol

```diff
@@ -163,17 +164,156 @@ contract VaultAdmin is VaultStorage {
+
+        // Moved to a separate function to avoid stack too deep errors
+        _postSwapChecks(
+            _fromAsset,
+            _fromAssetAmount,
+            fromAssetConfig,
+            _toAsset,
+            toAssetAmount,
+            _minToAssetAmount,
+            toAssetConfig,
+            config
+        );
+
+        emit Swapped(_fromAsset, _toAsset, _fromAssetAmount, toAssetAmount);
+    }
+
+    function _postSwapChecks(
+        address _fromAsset,
+        uint256 _fromAssetAmount,
+        Asset memory fromAssetConfig,
+        address _toAsset,
+        uint256 toAssetAmount,
+        uint256 _minToAssetAmount,
+        Asset memory toAssetConfig,
+        SwapConfig memory config
+    ) internal view {
+        // Check the to assets returned is above slippage amount specified by the strategist
+        require(
+            toAssetAmount >= _minToAssetAmount,
+            "Strategist slippage limit"
+        );
+
+        // Check the slippage against the Oracle in case the strategist made a mistake or has become malicious.
+        // to asset amount = from asset amount * from asset price / to asset price
+        uint256 minOracleToAssetAmount = (_fromAssetAmount *
+            (1e4 - fromAssetConfig.allowedOracleSlippageBps) *
+            IOracle(priceProvider).price(_fromAsset)) /
+            (IOracle(priceProvider).price(_toAsset) *
+                (1e4 + toAssetConfig.allowedOracleSlippageBps));
+
+        require(
+            toAssetAmount >= minOracleToAssetAmount,
+            "Oracle slippage limit exceeded"
+        );
+
+        // Check the vault's total value hasn't gone below the OToken total supply
+        // by more than the allowed percentage.
+        require(
+            (IVault(address(this)).totalValue() *
+                (1e4 + config.allowedUndervalueBps)) /
+                1e4 >=
+                oUSD.totalSupply(),
+            "Allowed value < supply"
+        );
+    }
+
+    /***************************************
+                    Swap Config
+    ****************************************/
+
+    /**
+     * @notice Set the contract the performs swaps of collateral assets.
      * @param _swapperAddr Address of the Swapper contract that implements the ISwapper interface.
      */
     function setSwapper(address _swapperAddr) external onlyGovernor {
-        swapper = _swapperAddr;
+        swapConfig.swapper = _swapperAddr;
         emit SwapperChanged(_swapperAddr);
     }
 
+    /// @notice Contract that swaps the vault's collateral assets
+    function swapper() external view returns (address swapper_) {
+        swapper_ = swapConfig.swapper;
+    }
+
+    /**
+     * @notice Set max allowed percentage the vault total value can drop below the OToken total supply in basis points
+     * when executing collateral swaps.
+     * @param _basis Percentage in basis points. eg 100 == 1%
+     */
+    function setSwapAllowedUndervalue(uint16 _basis) external onlyGovernor {
```

Comment:
Let's say we have a built in dripper, and have a reserve. This means that the resting vault value would be greater than the total supply by perhaps a few percentage points.

Because this config only works based off undervalue, we would be unable to specify that we want the vault value to remain greater than the total supply. 

We could change this to just be a direct multiplier onto the total supply, for example 10,000 = we need to keep the vault value at at least 100% of total supply after the swap. 10,100 would be 101%

Or we could just keep this as is. I'm okay either way.

Reply:
That's an interesting one. Wouldn't the vault's total value exclude assets that are still to be dripped into the vault? Sure, the assets are physically held in the vault, but I would have expected the total value function to be aware of unrealised dripped assets. So over time the total value slowly increases and any rebases will pick up the newly realised assets.

Regardless, you make a good point on the undervalue being based off the vault value. The term undervalue is referencing the total supply so I'll change the check to 
```Solidity
IVault(address(this)).totalValue() >=
    (oUSD.totalSupply() * ((1e4 - config.allowedUndervalueBps))) /
        1e4
```

I generally prefer multipliers over percentage/basis points as it uses slightly less gas. But I went for basis points here to be consistent with other vault config. eg `redeemFeeBps` and `trusteeFeeBps`